### PR TITLE
codemirror blob: First stab at hover cards

### PR DESCRIPTION
**Pull request by fkling**
Depends on #39448 

This is a first implementation of hovercards. It only renders content,
no actions and it doesn't support "pinning" cards. I tried to
approximate the current hovercard as much as possible but my main goal
was to get it working, not to make it look identical.
I gave the hovercards a bit more "space" but I still don't like how they
look like with few content.

I also add support for document highlights because this will also
highlight the token underneath the cursor. Unlike the current blob view
the token underneath the cursor isn't rendered in a different color (for
the time being).

I tested this with precise code intel information, I don't know how well
it works for search based information.

There are some issues wrt hovercard positioning and how quickly the cards
update, but those can be resolved in future PRs.

Demo: 

https://user-images.githubusercontent.com/179026/181229609-6405a05f-7ed1-4c81-a5c1-5ed3b0212410.mp4

## Test plan

Manual testing.

## App preview:

- [Web](https://sg-web-fkling-cm-file-popover.onrender.com)
- [Storybook](https://5f0f381c0e50750022dc6bf7-vmvcsicahn.chromatic.com)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.

### CODE REVIEW

File: client/web/src/repo/blob/ui/BlobStatusBarContainer.module.scss [hunk 1]
```diff
@@ -11,7 +11,6 @@
 }
 
 .content {
-    isolation: isolate;
```
Comment (fkling) [inline client/web/src/repo/blob/ui/BlobStatusBarContainer.module.scss:14]: If this is present, hoverards that extend outside of the blob view on the right hand side will be covered by the extensions sidebar, which is weird. I don't see why `isolate` is necessary here.

File: client/web/src/search/Notepad.module.scss [hunk 2]
```diff
@@ -1,4 +1,5 @@
 .root {
+    z-index: 100;
```
Comment (fkling) [inline client/web/src/search/Notepad.module.scss:2]: Ensures notepad is always on top.

**Review by olafurpg — APPROVED** [review]
I just tried this locally and it's very nice to see that both document highlights and hover tooltips working so well! LGTM 👍 I have some non-blocking thoughts that can be addressed separately

- Would it make sense to define max-width on the popover? It's difficult to read the text with such a wide box
CleanShot 2022-07-27 at 22 03 46@2x

For comparison, the current popover has a more narrow width
CleanShot 2022-07-27 at 22 06 09@2x

- What determines the position of the popover? It seems to switch between above and below the cursor for the same token. Reproduce in cmd/gitserver/server/vcs_syncer.go



- It might be worth adding a spinner when a hover request is triggered. Alternatively, display some hint in the UI that there is no hover available. I think it's worth exploring a less jumpy alternative to the current blob implementation where the popover appears for a second with a spinner and then disappears if there's no result. For example, we could something similar to `ProgressLocation.Window` from VS Code, which shows a flashing blue line at the top of the editor window

CleanShot 2022-07-27 at 22 14 46

https://sourcegraph.com/npm/types/vscode@43b010e8db7b69fb88730e90db3407480c91daf3/-/blob/index.d.ts?L9827:9&popover=pinned

**Comment by fkling** [comment]
Closing in favor of https://github.com/sourcegraph/sourcegraph/pull/39650

But to address @olafurpg's points:

> Would it make sense to define max-width on the popover? It's difficult to read the text with such a wide box

Uh yeah. I'm keep my browser window narrow so I didn't notice that. I find the current max-width to narrow but I can definitely set one.

> What determines the position of the popover? It seems to switch between above and below the cursor for the same token. 

CodeMirror does it somehow. That's one of the first things I noticed too and I have no idea what's going on. It could be that the card is not rendered yet when CodeMirror does measuring or it could be a bug with CodeMirror. I plan to investigate more but didn't want this to be a blocker.

>  I think it's worth exploring a less jumpy alternative to the current blob implementation where the popover appears for a second with a spinner and then disappears if there's no result.

That's why I'm ignoring loading state right now. Definitely something we should look into.

**Pull request closed without merging**